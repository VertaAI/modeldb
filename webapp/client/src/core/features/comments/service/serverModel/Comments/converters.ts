import { IComment } from '../../../Model';

export const convertServerComment = <Comment extends IComment>(
  serverComment: any
): Comment => {
  return {
    id: serverComment.id,
    dateTime: new Date(Number(serverComment.date_time)),
    message: serverComment.message,
  } as Comment;
};
