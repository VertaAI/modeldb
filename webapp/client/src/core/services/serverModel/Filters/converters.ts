import { IFilterData } from 'core/features/filter/Model';
import { getServerFilterValueType, getServerFilterOperator } from './Filters';

export interface IFilterConverter<
  R extends IServerFiltersInRequest,
  T extends IFilterData = IFilterData
> {
  predicate: (filter: IFilterData) => filter is T;
  convert: (filter: T, request: R) => Promise<R>;
}
export interface IServerFiltersInRequest {
  predicates?: any[];
}
export const makeFilterConverter = <
  R extends IServerFiltersInRequest,
  T extends IFilterData = IFilterData
>(
  filterConverter: IFilterConverter<R, T>
) => filterConverter;

const defaultFilterConverter = makeFilterConverter<IServerFiltersInRequest>({
  predicate: (_): _ is IFilterData => true,
  convert: (filter, request) => {
    return Promise.resolve({
      ...request,
      predicates: (request.predicates || []).concat({
        key: filter.name.toLocaleLowerCase() === 'tag' ? 'tags' : filter.name,
        value: Array.isArray(filter.value) ? filter.value[0] : filter.value,
        value_type: getServerFilterValueType(filter.type),
        operator: getServerFilterOperator(filter),
      }),
    });
  },
});

export const makeAddFiltersToRequest: <R extends IServerFiltersInRequest>(
  additionalFilterConverters?: Array<IFilterConverter<R, any>>
) => (filters: IFilterData[]) => (request: R) => Promise<R> = (
  additionalFilterConverters = []
) => filters => request => {
  const filterConverters = [
    ...additionalFilterConverters,
    defaultFilterConverter,
  ];

  return filters
    .map(filter => {
      const appropriateFilterConverter = filterConverters.find(
        ({ predicate }) => predicate(filter)
      );
      if (appropriateFilterConverter) {
        return [filter, appropriateFilterConverter.convert] as [
          IFilterData,
          IFilterConverter<any>['convert']
        ];
      }
      throw new Error(`not found converter for filter = ${filter}`);
    })
    .reduce((resPromise, [filter, convert]) => {
      return resPromise.then(resRequest => convert(filter, resRequest));
    }, Promise.resolve(request));
};
