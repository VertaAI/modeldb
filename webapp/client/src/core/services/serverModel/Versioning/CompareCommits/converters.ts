import * as R from 'ramda';

import { IBlob, blobCategories } from 'core/shared/models/Repository/Blob/Blob';
import { Diff, DiffType } from 'core/shared/models/Repository/Blob/Diff';
import { exhaustiveCheck } from 'core/shared/utils/exhaustiveCheck';
import matchType from 'core/shared/utils/matchType';

import { convertServerCodeDiff } from './convertServerCodeDiff';
import { convertServerConfigDiff } from './convertServerConfigDiff';
import convertServerDatasetDiff from './convertServerDatasetDiff';
import { convertServerEnvironmentDiff } from './convertServerEnvironmentDiff';

export const convertServerDiffsToClient = (serverDiff: any[]): Diff[] => {
  return (serverDiff || [])
    .filter((diff: any) => !R.isEmpty(diff))
    .map(convertServerDiffToClient);
};

export const convertServerDiffToClient = (serverDiff: any): Diff => {
  const serverDiffCategory = Object.keys(
    R.omit(['location', 'status'], serverDiff)
  )[0];
  const category: IBlob['data']['category'] =
    serverDiffCategory in blobCategories
      ? (serverDiffCategory as IBlob['data']['category'])
      : 'unknown';

  const diffType: DiffType = serverDiff.status
    ? matchType(
        {
          MODIFIED: () => 'updated',
          DELETED: () => 'deleted',
          ADDED: () => 'added',
        },
        serverDiff.status
      )
    : 'added';

  switch (category) {
    case 'code': {
      return convertServerCodeDiff(serverDiff, diffType);
    }

    case 'config': {
      return convertServerConfigDiff(serverDiff, diffType);
    }

    case 'environment': {
      return convertServerEnvironmentDiff(serverDiff, diffType);
    }

    case 'unknown': {
      switch (diffType) {
        case 'added':
        case 'deleted':
          return {
            diffType,
            category: 'unknown',
            type: 'unknown',
            location: serverDiff.location,
            blob: serverDiff[serverDiffCategory],
          };

        case 'updated': {
          return {
            diffType,
            category: 'unknown',
            type: 'unknown',
            location: serverDiff.location,
            blobA: serverDiff[serverDiffCategory].A,
            blobB: serverDiff[serverDiffCategory].B,
          };
        }

        default:
          exhaustiveCheck(diffType, '');
      }
    }

    case 'dataset': {
      return convertServerDatasetDiff(serverDiff, diffType);
    }

    default: {
      exhaustiveCheck(category, '');
    }
  }
};
