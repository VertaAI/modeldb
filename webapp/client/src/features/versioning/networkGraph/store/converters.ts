import {
  INetworkPoint,
  networkGraphSettings,
  INetworkConnection,
  INetworkData,
} from 'shared/models/Versioning/NetworkGraph';

import { Network_workspace_repository_network } from './graphql-types/Network';

export const convertNetwork = (
  data: Network_workspace_repository_network
): INetworkData => {
  const points: INetworkPoint[] = data.commits.map((commit, commitIndex) => {
    const branchIndex = data.branches.findIndex(b => b.color === commit.color);
    const branch = data.branches[branchIndex];

    return {
      branch: branch ? branch.branch : '',
      message: commit.commit.message || '',
      color: commit.color,
      isLastCommitInBranch: branch && branch.commitIndex === commitIndex,
      coordinate: {
        x:
          networkGraphSettings.startPaddingX +
          commitIndex * networkGraphSettings.xPointGap,
        y:
          networkGraphSettings.startPaddingY +
          commit.color * networkGraphSettings.yPointGap,
      },
      commitSha: commit.commit.sha,
      authorPicture: commit.commit.author.picture,
    };
  });

  const connections: INetworkConnection[] = data.edges.map(edge => {
    const fromPoint = points[edge.fromCommitIndex];
    const toPoint = points[edge.toCommitIndex];

    return {
      type: edge.edgeType as INetworkConnection['type'],
      color: edge.color,
      fromCoordinate: fromPoint.coordinate,
      toCoordinate: toPoint.coordinate,
    };
  });

  return {
    points,
    connections,
  };
};
