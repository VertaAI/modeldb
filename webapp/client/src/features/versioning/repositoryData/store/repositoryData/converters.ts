import {
  IBlobFolderElement,
  ISubFolderElement,
  IHydratedCommit,
} from 'shared/models/Versioning/RepositoryData';
import { convertServerBlobToClient } from 'services/serverModel/Versioning/RepositoryData/Blob';
import { convertUser } from 'shared/graphql/User/User';

import { ICommitComponentView } from '../types';
import {
  CommitWithComponent_repository_commitByReference_getLocation,
  CommitWithComponent_repository_commitByReference,
} from './graphql-types/CommitWithComponent';

export const convertCommitWithComponent = (
  serverCommit: CommitWithComponent_repository_commitByReference
): { commit: IHydratedCommit; component: ICommitComponentView } => {
  const component = convertComponent(serverCommit.getLocation);
  return {
    commit: {
      dateCreated: new Date(Number(serverCommit.date)),
      message: serverCommit.message,
      sha: serverCommit.id,
      author: convertUser(serverCommit.author),
      type: component ? 'withParent' : 'initial',
    },
    component,
  };
};

export const convertComponent = (
  serverCommitComponent: CommitWithComponent_repository_commitByReference_getLocation | null
): ICommitComponentView => {
  if (!serverCommitComponent) {
    return { type: 'folder', blobs: [], subFolders: [] };
  }
  switch (serverCommitComponent.__typename) {
    case 'CommitBlob': {
      const blob = convertServerBlobToClient(
        JSON.parse(serverCommitComponent.content)
      );
      return {
        type: 'blob',
        data: blob.data,
        experimentRuns: serverCommitComponent.runs.runs,
      };
    }
    case 'CommitFolder': {
      return {
        type: 'folder',
        blobs: serverCommitComponent.blobs.map(({ name }) => {
          const res: IBlobFolderElement = {
            type: 'blob',
            name: name as IBlobFolderElement['name'],
          };
          return res;
        }),
        subFolders: serverCommitComponent.subfolders.map(({ name }) => {
          const res: ISubFolderElement = {
            type: 'folder',
            name: name as ISubFolderElement['name'],
          };
          return res;
        }),
      };
    }
  }
};
