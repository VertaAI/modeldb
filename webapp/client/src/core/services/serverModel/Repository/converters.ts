import { IRepository, Label } from 'core/shared/models/Repository/Repository';
import { convertServerEntityWithLoggedDates } from 'services/serverModel/Common/converters';
import { convertServerShortWorkspaceToClient } from 'services/serverModel/Workspace/converters';

import User from 'models/User';
import { IServerRepository } from './Repository';

export const convertServerRepositoryToClient = ({
  serverRepository,
  labels,
  owner,
}: {
  serverRepository: IServerRepository;
  labels: Label[];
  owner: User;
}): IRepository => {
  const { dateCreated, dateUpdated } = convertServerEntityWithLoggedDates({
    date_created: serverRepository.date_created,
    date_updated: serverRepository.date_updated,
  });
  return {
    id: serverRepository.id,
    name: serverRepository.name,
    shortWorkspace: convertServerShortWorkspaceToClient(serverRepository),
    dateCreated,
    dateUpdated,
    labels,
    owner,
  };
};
