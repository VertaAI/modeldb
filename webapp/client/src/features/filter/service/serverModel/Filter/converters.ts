import { makeAddFiltersToRequest } from 'core/features/filter/service/serverModel/Filters/converters';

export const makeAddFiltersToRequestWithDefaultFilters = makeAddFiltersToRequest;
