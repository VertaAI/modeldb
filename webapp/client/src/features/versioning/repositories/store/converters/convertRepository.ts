import { IRepository } from 'shared/models/Versioning/Repository';
import { ICurrentWorkspace } from 'shared/models/Workspace';

import { Repository_workspace_repository } from '../repositoryQuery/graphql-types/Repository';
import { convertUser } from 'shared/graphql/User/User';

export const convertRepository = (
  serverRepository: Repository_workspace_repository,
  currentWorkspace: ICurrentWorkspace
) => {
  const repository: IRepository = {
    dateCreated: new Date(Number(serverRepository.dateCreated)),
    dateUpdated: new Date(Number(serverRepository.dateUpdated)),
    id: serverRepository.id,
    labels: serverRepository.labels,
    name: serverRepository.name,
    owner: convertUser(serverRepository.owner),
    shortWorkspace: currentWorkspace as IRepository['shortWorkspace'],
  };
  return repository;
};
