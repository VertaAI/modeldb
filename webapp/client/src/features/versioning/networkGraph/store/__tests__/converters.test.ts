import { convertNetwork } from '../converters';
import { networkHistoryMock } from './__mocks__/networkHistoryMock';

describe('(networkGraph feature)', () => {
  describe('(networkGraph converters', () => {
    it('"convertNetworkHistory" should convert data correctly', () => {
      expect(convertNetwork(networkHistoryMock)).toMatchSnapshot();
    });
  });
});
