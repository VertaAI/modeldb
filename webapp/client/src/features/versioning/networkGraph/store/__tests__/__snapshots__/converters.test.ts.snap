// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(networkGraph feature) (networkGraph converters "convertNetworkHistory" should convert data correctly 1`] = `
Object {
  "connections": Array [
    Object {
      "color": 0,
      "fromCoordinate": Object {
        "x": 20,
        "y": 20,
      },
      "toCoordinate": Object {
        "x": 40,
        "y": 20,
      },
      "type": "DEFAULT",
    },
    Object {
      "color": 1,
      "fromCoordinate": Object {
        "x": 40,
        "y": 20,
      },
      "toCoordinate": Object {
        "x": 60,
        "y": 40,
      },
      "type": "BRANCH",
    },
    Object {
      "color": 1,
      "fromCoordinate": Object {
        "x": 60,
        "y": 40,
      },
      "toCoordinate": Object {
        "x": 80,
        "y": 20,
      },
      "type": "MERGE",
    },
    Object {
      "color": 2,
      "fromCoordinate": Object {
        "x": 60,
        "y": 40,
      },
      "toCoordinate": Object {
        "x": 100,
        "y": 60,
      },
      "type": "BRANCH",
    },
    Object {
      "color": 2,
      "fromCoordinate": Object {
        "x": 100,
        "y": 60,
      },
      "toCoordinate": Object {
        "x": 120,
        "y": 20,
      },
      "type": "MERGE",
    },
    Object {
      "color": 0,
      "fromCoordinate": Object {
        "x": 40,
        "y": 20,
      },
      "toCoordinate": Object {
        "x": 80,
        "y": 20,
      },
      "type": "DEFAULT",
    },
    Object {
      "color": 0,
      "fromCoordinate": Object {
        "x": 80,
        "y": 20,
      },
      "toCoordinate": Object {
        "x": 120,
        "y": 20,
      },
      "type": "DEFAULT",
    },
  ],
  "points": Array [
    Object {
      "authorPicture": null,
      "branch": "b1",
      "color": 0,
      "commitSha": "ee4ce00cfc5406da3934a347df9c045bd75371e8eb1bd5fb68ee3fa30303bd76",
      "coordinate": Object {
        "x": 20,
        "y": 20,
      },
      "isLastCommitInBranch": false,
      "message": "Test message",
    },
    Object {
      "authorPicture": null,
      "branch": "b1",
      "color": 0,
      "commitSha": "c44b83ff69b1eb794cb1eedae59d9d74dc4db8afac8a45abc9f93bda324d45e9",
      "coordinate": Object {
        "x": 40,
        "y": 20,
      },
      "isLastCommitInBranch": false,
      "message": "Test message",
    },
    Object {
      "authorPicture": null,
      "branch": "b2",
      "color": 1,
      "commitSha": "eeca7178617b8f8c7d8f6584745dbdc965dfa0141228e03c5f442b98dedd0150",
      "coordinate": Object {
        "x": 60,
        "y": 40,
      },
      "isLastCommitInBranch": true,
      "message": "Test message",
    },
    Object {
      "authorPicture": null,
      "branch": "b1",
      "color": 0,
      "commitSha": "846ecb717649ce258b37370b529e51a8616263dc0e8356fa62c59b83bb35e9ad",
      "coordinate": Object {
        "x": 80,
        "y": 20,
      },
      "isLastCommitInBranch": false,
      "message": "Test message",
    },
    Object {
      "authorPicture": null,
      "branch": "b3",
      "color": 2,
      "commitSha": "b3516e6fe3451684abb4b3b8cf488e74673de09ef7ebd7a9fe512e7d6badf15e",
      "coordinate": Object {
        "x": 100,
        "y": 60,
      },
      "isLastCommitInBranch": true,
      "message": "Test message",
    },
    Object {
      "authorPicture": null,
      "branch": "b1",
      "color": 0,
      "commitSha": "b3516e6fe3451684abb4b3b8cf488e74673de09ef7ebd7a9fe512e7d6badf15e",
      "coordinate": Object {
        "x": 120,
        "y": 20,
      },
      "isLastCommitInBranch": true,
      "message": "Test message",
    },
  ],
}
`;
